File: AN_2/SEM_2/MPP/java/lab5/lab5/Server/src/main/java/ro/mpp2024/server/ServicesImpl.java
package ro.mpp2024.server;

import ro.mpp2024.model.*;
import ro.mpp2024.persistence.*;
import ro.mpp2024.services.*;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

public class ServicesImpl implements IServices {

    private CursaRepositoryInterface cursaRepository;
    private EchipaRepositoryInterface echipaRepository;
    private ParticipantRepositoryInterface participantRepository;
    private ParticipantCursaRepositoryInterface participantCursaRepository;
    private UtilizatorRepositoryInterface utilizatorRepository;

    private Map<String, IObserver> loggedObservers;

    public ServicesImpl(UtilizatorRepositoryInterface utilizatorRepository, EchipaRepositoryInterface echipaRepository, ParticipantRepositoryInterface participantRepository, CursaRepositoryInterface cursaRepository, ParticipantCursaRepositoryInterface participantCursaRepository) {
        this.utilizatorRepository = utilizatorRepository;
        this.echipaRepository = echipaRepository;
        this.participantRepository = participantRepository;
        this.cursaRepository = cursaRepository;
        this.participantCursaRepository = participantCursaRepository;
        loggedObservers = new ConcurrentHashMap<>();
    }

    @Override
    public synchronized Utilizator login(String nume, String parola, IObserver client) throws Exception {
        Utilizator utilizator = utilizatorRepository.findByNume(nume);
        if (loggedObservers.containsKey(nume)) {
            throw new Exception("Utilizatorul este deja conectat.");
        }
        if (utilizator != null && utilizator.verificaParola(parola)) {
            loggedObservers.put(nume, client);
            return utilizator;
        }
        throw new Exception("Autentificare esuata.");
    }

    @Override
    public synchronized void inscriereParticipant(Participant participant) throws Exception {
        participantRepository.adauga(participant);
    }

    @Override
    public synchronized List<Participant> findAllParticipanti() throws Exception {
        return participantRepository.findAll();
    }

    @Override
    public synchronized List<Participant> cautaParticipantiDupaEchipa(String numeEchipa) throws Exception {
        Echipa echipa = echipaRepository.findByNume(numeEchipa);
        if (echipa == null) {
            throw new Exception("Echipa nu exista");
        }
        return participantRepository.findByEchipa(echipa);
    }


    @Override
    public synchronized Echipa findEchipaByNume(String nume) throws Exception {
        return echipaRepository.findByNume(nume);
    }

    @Override
    public synchronized void adaugaEchipa(Echipa echipa) throws Exception {
        echipaRepository.adauga(echipa);
    }

    @Override
    public synchronized List<Cursa> findAllCurse() throws Exception {
        return cursaRepository.findAll();
    }

    @Override
    public synchronized List<CursaDTO> listDtoCurse() throws Exception {
        List<Cursa> curse = cursaRepository.findAll();
        List<CursaDTO> dtoList = new ArrayList<>();
        for (Cursa cursa : curse) {
            int nrPart = participantCursaRepository.findNrPartByCursa(cursa);
            dtoList.add(new CursaDTO(cursa.getNume(), cursa.getCapacitateMinima(), cursa.getCapacitateMaxima(), nrPart));
        }
        return dtoList;
    }

    @Override
    public synchronized int findNrPartByCursa(Cursa cursa) throws Exception {
        return participantCursaRepository.findNrPartByCursa(cursa);
    }

    private void notifyParticipantAdded(Participant participant, Cursa cursa) {
        for (IObserver obs : loggedObservers.values()) {
            try {
                obs.adauga(participant, cursa);
            } catch (Exception e) {
                System.out.println("Error notifying observer: " + e.getMessage());
            }
        }
    }

    @Override
    public synchronized void adaugaParticipantLaCursa(Participant participant, Cursa cursa) throws Exception {
        participantCursaRepository.adaugainbaza(participant, cursa);
        notifyParticipantAdded(participant, cursa);
    }

    @Override
    public synchronized List<ParticipantCursa> findAllParticipantCursa() throws Exception {
        return participantCursaRepository.findAll();
    }

}

File: AN_2/SEM_2/MPP/c#/lab6/Server/mpp.server/ServicesImpl.cs
using System.Collections.Concurrent;

namespace Server.mpp.server;
using Model.mpp.model;
using Persistence.mpp.persistence;
using Services.mpp.services;

public class ServicesImpl : IServices
{
    private readonly ICursaRepository _cursaRepository;
    private readonly IEchipaRepository _echipaRepository;
    private readonly IParticipantRepository _participantRepository;
    private readonly IParticipantCursaRepository _participantCursaRepository;
    private readonly IUtilizatorRepository _utilizatorRepository;
    private readonly ConcurrentDictionary<string, List<IObserver>> loggedClients = new();

    public ServicesImpl(
        IUtilizatorRepository utilizatorRepository,
        IEchipaRepository echipaRepository,
        IParticipantRepository participantRepository,
        ICursaRepository cursaRepository,
        IParticipantCursaRepository participantCursaRepository)
    {
        _utilizatorRepository = utilizatorRepository;
        _echipaRepository = echipaRepository;
        _participantRepository = participantRepository;
        _cursaRepository = cursaRepository;
        _participantCursaRepository = participantCursaRepository;
    }

    public void Login(Utilizator utilizator, IObserver client)
    {
        if (utilizator == null)
            throw new Exception("Utilizator invalid.");

        lock (loggedClients)
        {
            if (!loggedClients.ContainsKey(utilizator.Nume))
            {
                loggedClients[utilizator.Nume] = new List<IObserver>();
            }

            var observers = loggedClients[utilizator.Nume];

            if (!observers.Contains(client))
            {
                observers.Add(client);
            }
            else
            {
                throw new Exception("Observer deja logat.");
            }
        }
    }
    
    public void Logout(Utilizator utilizator, IObserver client)
    {
        if (utilizator == null)
            return;

        lock (loggedClients)
        {
            if (loggedClients.TryGetValue(utilizator.Nume, out var observers))
            {
                observers.Remove(client);

                if (observers.Count == 0)
                {
                    loggedClients.TryRemove(utilizator.Nume, out _);
                }
            }
        }

        Console.WriteLine($"{utilizator.Nume} s-a delogat (observer: {client.GetType().Name})");
    }

    public Participant FindByNume(string nume)
    {
        return _participantRepository.FindByNume(nume);
    }
    
    public void InscriereParticipant(Participant participant)
    {
        _participantRepository.Adauga(participant);
    }

    public List<Participant> FindAllParticipanti()
    {
        return _participantRepository.FindAll();
    }

    public List<Participant> CautaParticipantiDupaEchipa(string numeEchipa)
    {
        var echipa = _echipaRepository.FindByNume(numeEchipa);
        if (echipa == null)
            throw new Exception("Echipa nu exista");

        return _participantRepository.FindByEchipa(echipa);
    }

    public Echipa FindEchipaByNume(string nume)
    {
        return _echipaRepository.FindByNume(nume);
    }

    public void AdaugaEchipa(Echipa echipa)
    {
        _echipaRepository.Adauga(echipa);
    }

    public List<Cursa> FindAllCurse()
    {
        return _cursaRepository.FindAll();
    }

    public List<CursaDTO> ListDtoCurse()
    {
        var curse = _cursaRepository.FindAll();
        var result = new List<CursaDTO>();
        foreach (var cursa in curse)
        {
            int nrPart = _participantCursaRepository.CountByCursa(cursa);
            result.Add(new CursaDTO(cursa.Nume, cursa.CapacitateMinima, cursa.CapacitateMaxima, nrPart));
        }
        return result;
    }


    public int FindNrPartByCursa(Cursa cursa)
    {
        return _participantCursaRepository.CountByCursa(cursa);
    }

    public void AdaugaParticipantLaCursa(ParticipantCursa participantCursa)
    {
        var participantComplet = _participantRepository.FindById(participantCursa.Participant.Id);

        if (participantComplet == null)
        {
            throw new Exception("Participantul nu a fost găsit!");
        }

        var participantCursaComplet = new ParticipantCursa(participantComplet, participantCursa.Cursa);

        _participantCursaRepository.adaugainbaza(participantCursaComplet);

        NotifyParticipantAdded(participantCursaComplet);
    }

    public List<ParticipantCursa> FindAllParticipantCursa()
    {
        return _participantCursaRepository.FindAll();
    }

    private void NotifyParticipantAdded(ParticipantCursa participantCursa)
    {
        if (participantCursa == null)
            return;

        foreach (var observers in loggedClients.Values)
        {
            foreach (var observer in observers)
            {
                Task.Run(async () =>
                {
                    try
                    {
                        await Task.Delay(10); 
                        observer.Adauga(participantCursa);
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine("Error notifying observer: " + ex.Message);
                    }
                });
            }
        }
    }
}